src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp

```diff
@@ -87,29 +87,215 @@ void ShenandoahDirectCardMarkRememberedSet::merge_overreach(size_t first_cluster
+  size_t num_groups = 0;
+  size_t cumulative_group_span = 0;
+  size_t current_group_span = _first_group_chunk_size * _group_size;
+  size_t smallest_group_span = _smallest_chunk_size * _group_size;
+  while (unspanned_heap_size > 0) {
+    if (current_group_span <= unspanned_heap_size) {
+      unspanned_heap_size -= current_group_span;
+      num_chunks += _group_size;
+      num_groups++;
+
+      if (num_groups >= _num_groups) {
+        // The last group has more than _group_size entries.
+        size_t chunk_span = current_group_span / _group_size;
+        size_t extra_chunks = unspanned_heap_size / chunk_span;
+        assert (extra_chunks * chunk_span == unspanned_heap_size, "Chunks must precisely span regions");
+        num_chunks += extra_chunks;
+        return num_chunks;
+      } else if (current_group_span <= smallest_group_span) {
+        // We cannot introduce new groups because we've reached the lower bound on group size
+        size_t chunk_span = _smallest_chunk_size;
+        size_t extra_chunks = unspanned_heap_size / chunk_span;
+        assert (extra_chunks * chunk_span == unspanned_heap_size, "Chunks must precisely span regions");
+        num_chunks += extra_chunks;
+        return num_chunks;
+      } else {
+        current_group_span /= 2;
+      }
+    } else {
+      // The last group has fewer than _group_size entries.
+      size_t chunk_span = current_group_span / _group_size;
+      size_t last_group_size = unspanned_heap_size / chunk_span;
+      assert (last_group_size * chunk_span == unspanned_heap_size, "Chunks must precisely span regions");
+      num_chunks += last_group_size;
+      return num_chunks;
     }
-    region = _regions->next();
   }
+  return num_chunks;
+}
+
+ShenandoahRegionChunkIterator::ShenandoahRegionChunkIterator(size_t worker_count) :
+    _heap(ShenandoahHeap::heap()),
+    _group_size(calc_group_size()),
+    _first_group_chunk_size(calc_first_group_chunk_size()),
+    _num_groups(calc_num_groups()),
+    _total_chunks(calc_total_chunks()),
+    _index(0)
+{
+  assert(_smallest_chunk_size ==
+         CardTable::card_size_in_words() * ShenandoahCardCluster<ShenandoahDirectCardMarkRememberedSet>::CardsPerCluster,
+         "_smallest_chunk_size is not valid");
+
+  size_t words_in_region = ShenandoahHeapRegion::region_size_words();
+  size_t group_span = _first_group_chunk_size * _group_size;
+
+  _region_index[0] = 0;
+  _group_offset[0] = 0;
+  for (size_t i = 1; i < _num_groups; i++) {
+    _region_index[i] = _region_index[i-1] + (_group_offset[i-1] + group_span) / words_in_region;
+    _group_offset[i] = (_group_offset[i-1] + group_span) % words_in_region;
+    group_span /= 2;
+  }
+  // Not necessary, but keeps things tidy
+  for (size_t i = _num_groups; i < _maximum_groups; i++) {
+    _region_index[i] = 0;
+    _group_offset[i] = 0;
+  }
+}
+
+ShenandoahRegionChunkIterator::ShenandoahRegionChunkIterator(ShenandoahHeap* heap, size_t worker_count) :
```

Comment:
Can we use [constructor delegation](https://en.wikipedia.org/wiki/C++11#Object_construction_improvement) here to remove some duplicated code? Or could we just have clients always pass in the heap reference? One constructor is missing an assert.

Reply:
Thanks for this suggestion.  I've removed body of one constructor and replaced with constructor delegation.